### PR TITLE
squash: add -A/-B/-d options

### PR DESCRIPTION
**Pull request by glehmann**
# Checklist

If applicable:

- [x] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (`README.md`, `docs/`, `demos/`)
- [ ] I have updated the config schema (`cli/src/config-schema.json`)
- [x] I have added/updated tests to cover my changes

### CODE REVIEW

**Comment by bnjmnt4n** [comment]
How would this differ from `jj split`?

**Comment by glehmann** [comment]
> How would this differ from `jj split`?

`jj split` and `jj squash` are indeed quite close when using a single source commit, especially when used with a fileset or `-i`, and with one of the `-A/-B/-d` flag.

The difference comes when using multiple source commits. For example, with a commit tree like this one:

```
@  kkrvspxk (empty) (no description set)
○  lpllnppl file4
│  A file4
○  xynqvmyt file3
│  A file3
○  wkrztqwl file2
│  A file2
○  mzuowqnz file1
│  A file1
♦  zzzzzzzz root()
```

we can `jj squash -f x:: -d m` to squash `xynqvmyt` and all its descendant into a new commit onto `mzuowqnz`.

```
@  youptqqn (empty) (no description set)
○  wkrztqwl file2
│  A file2
│ ○  vsonsouy file3 file4
├─╯  A file3
│    A file4
○  mzuowqnz file1
│  A file1
♦  zzzzzzzz root()
```

IMO this PR and an implementation of #5301 would greatly improve the flexibility of `jj squash`.

**Comment by bnjmnt4n** [comment]
Thanks for the reminder, I completely forgot that `jj squash` accepted multiple source revsets 🤦

File: cli/src/commands/squash.rs [hunk 1]
```diff
@@ -89,30 +95,79 @@ pub(crate) struct SquashArgs {
...
     into: Option<RevisionArg>,
+
+    /// The revision(s) to use as parent for the new commit (can be repeated
+    /// to create a merge commit)
+    #[arg(
+        long,
+        short,
+        conflicts_with = "into",
+        conflicts_with = "revision",
+        conflicts_with = "use_destination_message",
```
Comment (yuja) [inline cli/src/commands/squash.rs]: nit: I think `--use-destination-message` can be allowed. It's basically an option to not open editor.

File: cli/src/commands/squash.rs [hunk 2]
```diff
@@ -124,22 +179,34 @@ pub(crate) fn cmd_squash(
...
-        // remove the destination from the sources
-        sources.retain(|source| source.id() != destination.id());
+        if use_new_flags {
+            pre_existing_destination = None;
+        } else {
+            let destination = if let Some(into) = &args.into {
+                workspace_command.resolve_single_rev(ui, into)?
+            } else {
+                workspace_command.resolve_single_rev(ui, &RevisionArg::AT)?
+            };
```
Comment (yuja) [inline cli/src/commands/squash.rs]: nit: unrelated change? `.as_ref().unwrap_or()` seems good.

File: cli/src/commands/squash.rs [hunk 3]
```diff
@@ -124,22 +179,34 @@ pub(crate) fn cmd_squash(
     command: &CommandHelper,
     args: &SquashArgs,
 ) -> Result<(), CommandError> {
+    let use_new_flags =
+        args.destination.is_some() || args.insert_after.is_some() || args.insert_before.is_some();
```
Comment (yuja) [inline cli/src/commands/squash.rs]: nit: `insert_destination_commit` or something might be better. "new flags" sounds like `--into`/`-r` are legacy options.

File: cli/src/commands/squash.rs [hunk 4]
```diff
@@ -155,21 +222,106 @@ pub(crate) fn cmd_squash(
...
+        let (parent_ids, child_ids) = compute_commit_location(
+            ui,
+            tx.base_workspace_helper(),
+            args.destination.as_deref(),
+            args.insert_after.as_deref(),
+            args.insert_before.as_deref(),
+            "squash commit",
+        )?;
+        tx.base_workspace_helper()
+            .check_rewritable(sources.iter().ids().chain(child_ids.iter()))?;
```
Comment (yuja) [inline cli/src/commands/squash.rs]: `sources` are tested before, and `child_ids` appears to be checked by `compute_commit_locations()`.

Can you also update cli/tests/test_immutable_commits.rs?

File: cli/src/commands/squash.rs [hunk 5]
```diff
@@ -155,21 +222,106 @@ pub(crate) fn cmd_squash(
...
+        let mut commit_builder = tx
+            .repo_mut()
+            .new_commit(parent_ids.clone(), merged_tree.id())
+            .detach();
+        let mut description = join_message_paragraphs(&args.message_paragraphs);
+        if !description.is_empty() {
+            commit_builder.set_description(description);
+            description = add_trailers(ui, &tx, &commit_builder)?;
+        }
+        commit_builder.set_description(&description);
```
Comment (yuja) [inline cli/src/commands/squash.rs]: I think description can be empty here, but I haven't tried.

File: cli/src/commands/squash.rs [hunk 6]
```diff
@@ -155,21 +222,106 @@ pub(crate) fn cmd_squash(
...
+            let child_commit = tx.repo().store().get_commit(child_id)?;
+            let new_parent_ids = child_commit
+                .parent_ids()
+                .iter()
+                .filter(|id| !&parent_ids.contains(id))
+                .cloned()
+                .chain(std::iter::once(commit.id().clone()))
+                .collect_vec();
+            let new_child_commit =
+                rebase_commit(tx.repo_mut(), child_commit, new_parent_ids).block_on()?;
```
Comment (yuja) [inline cli/src/commands/squash.rs]: Maybe we should do rebase in `.transform_descendats()`.

I think we can use `duplicate_commits(.., &[root], ..)` to insert an empty placeholder commit, but there might be a better function.

File: cli/src/commands/squash.rs [hunk 7]
```diff
@@ -155,21 +222,106 @@ pub(crate) fn cmd_squash(
...
+                if let Some(source_commit) = sources
+                    .iter()
+                    .position(|source| source.id() == &old_commit_id)
+                    .and_then(|index| sources.get_mut(index))
+                {
+                    *source_commit = new_commit;
+                }
+                Ok(())
+            })?;
+        commit
```
Comment (yuja) [inline cli/src/commands/squash.rs:284]: nit: we'll need to print the number of the rebased commits at the end of the operation.

File: cli/src/commands/squash.rs (hunk 2 above)
Comment (glehmann) [reply cli/src/commands/squash.rs]: The code has evolved on my side, to settle on this—code evolution has something organic :-)

I'll restore the previous version.

File: cli/src/commands/squash.rs (hunk 5 above)
Comment (glehmann) [reply cli/src/commands/squash.rs]: I think you're right. I'll give it a try

File: cli/src/commands/squash.rs (hunk 7 above)
Comment (glehmann) [reply cli/src/commands/squash.rs:284]: I think we are already good as is: the destination commit will be modified with the changes to squash. As a consequence, all its descendants will be counted as rebased.
There are effectively some intermediate rebases that are not counted, but I don't think we want to show that to the user.

File: cli/src/commands/squash.rs (hunk 6 above)
Comment (glehmann) [reply cli/src/commands/squash.rs]: We still need to update the sources with the new commits if some of the sources have been rewritten.
I've made the code more concise.

File: cli/src/commands/squash.rs [hunk 8]
```diff
@@ -155,21 +217,80 @@ pub(crate) fn cmd_squash(
...
+    } else {
+        // create the new destination commit
+        let (parent_ids, child_ids) = compute_commit_location(
+            ui,
+            tx.base_workspace_helper(),
+            args.destination.as_deref(),
+            args.insert_after.as_deref(),
+            args.insert_before.as_deref(),
+            "squash commit",
```
Comment (yuja) [inline cli/src/commands/squash.rs]: nit: "squashed commit" (past tense) or "rebased commit". I'm not sure which one is more correct.

File: cli/src/commands/squash.rs [hunk 9]
```diff
@@ -155,21 +217,80 @@ pub(crate) fn cmd_squash(
...
+            })
+            .try_collect()?;
+        let merged_tree = merge_commit_trees(tx.repo(), &parent_commits).block_on()?;
+        let commit = tx
+            .repo_mut()
+            .new_commit(parent_ids.clone(), merged_tree.id())
+            .write()?;
+        let mut rewritten = HashMap::new();
+        tx.repo_mut()
+            .transform_descendants(child_ids.clone(), async |mut rewriter| {
```
Comment (yuja) [inline cli/src/commands/squash.rs]: nit: can you update cli/tests/test_immutable_commits.rs?

File: cli/src/commands/squash.rs [hunk 10]
```diff
@@ -155,21 +217,80 @@ pub(crate) fn cmd_squash(
...
-    let matcher = workspace_command
+    workspace_command.check_rewritable(sources.iter().chain(&pre_existing_destination).ids())?;
+
+    // prepare the tx description before possibly rebasing the source commits
+    let source_ids: Vec<_> = sources.iter().ids().collect();
+    let tx_description = if let Some(destination) = &pre_existing_destination {
+        format!("squash commits into {}", destination.id().hex())
+    } else {
+        match &source_ids[..] {
+            [] => format!("squash {} commits", source_ids.len()),
```
Comment (yuja) [inline cli/src/commands/squash.rs:231]: Can you add test for empty source revision and tree?

I don't have strong feeling whether new commit should be inserted in that case, but the result and status output should be reasonable.

File: cli/src/commands/squash.rs [hunk 11]
```diff
@@ -197,7 +318,10 @@ pub(crate) fn cmd_squash(
             }
             SquashedDescription::Combine => {
                 let abandoned_commits = &squashed.abandoned_commits;
-                if let Some(description) = try_combine_messages(abandoned_commits, &destination) {
+                // TODO: chain the let expression, once our MSRV is at least 1.88
+                let description = try_combine_messages(abandoned_commits, &destination);
+                if !insert_destination_commit && description.is_some() {
```
Comment (yuja) [inline cli/src/commands/squash.rs]: Shouldn't we use the combined message in all cases?

Can you add test for that case, `--use-destination -d`, etc.?

File: cli/src/commands/squash.rs [hunk 12]
```diff
@@ -209,7 +333,11 @@ pub(crate) fn cmd_squash(
                         ui,
                         &tx,
                         abandoned_commits,
-                        &destination,
+                        if insert_destination_commit {
+                            None
+                        } else {
+                            Some(&destination)
+                        },
```
Comment (yuja) [inline cli/src/commands/squash.rs]: nit: (at your option) `(!insert_destination_commit).then_some(&destination)`

File: cli/tests/test_squash_command.rs [hunk 13]
```diff
@@ -1614,6 +1614,425 @@ fn test_squash_use_destination_message_and_message_mutual_exclusion() {
     [EOF]
     [exit status: 2]
     ");
+
+    insta::assert_snapshot!(work_dir.run_jj([
+        "squash",
+        "-r@",
+        "--into",
+        "@-"
```
Comment (yuja) [inline cli/tests/test_squash_command.rs]: nit: `--into=@-` to save vertical space?

**Review by yuja — APPROVED** [review]
Thanks.

File: cli/src/commands/squash.rs (hunk 9 above)
Comment (glehmann) [reply cli/src/commands/squash.rs]: I was quite sure it was already done (because you already asked for it), but I must have lost it somehow. Sorry about that.

File: cli/src/commands/squash.rs (hunk 12 above)
Comment (glehmann) [reply cli/src/commands/squash.rs]: nice!

File: cli/src/commands/squash.rs (hunk 11 above)
Comment (glehmann) [reply cli/src/commands/squash.rs]: > Shouldn't we use the combined message in all cases?

That felt obvious when I did it, but now I can't find why we wouldn't want to do that for a new destination…

I'll remove it.

File: cli/src/commands/squash.rs (hunk 7 above)
Comment (glehmann) [reply cli/src/commands/squash.rs:284]: We actually need to print the number of rebased commits when there are no changes to merge into the new destination

File: cli/src/commands/squash.rs (hunk 10 above)
Comment (glehmann) [reply cli/src/commands/squash.rs:231]: I ended up adding a new status
```
Created new commit xlzxqlsl 8ceb6c68 file 3&4
```
I'm not sure if it should be more specific

**Comment by ilyagr** [comment]
**Update:** I alluded to this below, but it's worth putting it first: this is a cool feature, thanks for working on it! I will certainly try it out.

I have a few worries about this.

- I think we need to document it better. Currently, it's quite hard to find out what `-d` does from jj's docs. Gaetan, would you be willing to do it, or should I try?
- I don't think this fits very well with what the current `jj squash` does, and I don't like reusing `--from` for this. I wonder if it'd be better to separate the two commands. If we call `newsplit` the behavior in this PR (we'd clearly need a better name), I think it'd make more sense to have `jj newsquah -r a -r b -d c` instead of `jj newsquah --from a --from b -d c`. Also, I think that if a revset has a single root, `jj newsquash -r revset` should collapse the revset into its root. (I think @martinvonz once suggested it).
- I'd also like to rename this command because, in some of its versions, it's closer to `split` than to `squash`. I don't have a cool name in mind right now, though. **Update:** Maybe `jj mix`? You pick out pieces of (or all of) one or more commits, mix them together in a bowl, and put the resulting commit somewhere.

At the same time, I think it's a cool idea and a cool experiment.

If people agree with these worries, I'd recommend marking the new options as "EXPERIMENTAL" until the worries are resolved (we have docs and we have decided on the UI).

**Comment by glehmann** [comment]
> **Update:** I alluded to this below, but it's worth putting it first: this is a cool feature, thanks for working on it! I will certainly try it out.

Thanks for looking into this!

> I have a few worries about this.
> 
>     * I think we need to document it better. Currently, it's quite hard to find out what `-d` does from jj's docs. Gaetan, would you be willing to do it, or should I try?

As I wrote in https://github.com/jj-vcs/jj/issues/7205#issuecomment-3209433493, I only see it as a commodity feature, same as we added in split. OTOH, I haven't even said that in the documentation.
But I have the feeling we are approaching the squash command in a significantly different way, otherwise I would probably understand why this would need more documentation than for split, or the feature would just make sense to you.

>     * I don't think this fits very well with what the current `jj squash` does, and I don't like reusing `--from` for this.

I don't see it as reusing for a different feature—again I think we see squash quite differently, but I'm not yet sure how.
For me `jj squash -r x` is just a shortcut for `jj squash -f x -t x-`.
`jj squash -f x -t z` is how we squash `x` into `z`.
`jj squash -f x::y -t z` is how we squash multiple revisions `x::y` into `z`.
`jj squash -f x::y -d z` is the same as above but squashing `x::y` in a new revision on top of `z`.

> I wonder if it'd be better to separate the two commands. If we call `newsplit` the behavior in this PR (we'd clearly need a better name), I think it'd make more sense to have `jj newsquah -r a -r b -d c` instead of `jj newsquah --from a --from b -d c`. Also, I think that if a revset has a single root, `jj newsquash -r revset` should collapse the revset into its root. (I think @martinvonz once suggested it).

I would definitely go that way for `-r` too. I've seen it in #5301.
But I'm not sure we need a new command for that.

>     * I'd also like to rename this command because, in some of its versions, it's closer to `split` than to `squash`. I don't have a cool name in mind right now, though. **Update:** Maybe `jj mix`? You pick out pieces of (or all of) one or more commits, mix them together in a bowl, and put the resulting commit somewhere.
> 
> 
> At the same time, I think it's a cool idea and a cool experiment.
> 
> If people agree with these worries, I'd recommend marking the new options as "EXPERIMENTAL" until the worries are resolved (we have docs and we have decided on the UI).

**Comment by ilyagr** [comment]
> But I have the feeling we are approaching the squash command in a significantly different way, otherwise I would probably understand why this would need more documentation than for split, or the feature would just make sense to you.

I can see the point of view from which this is a natural extension. It seems correct, but not obviously correct. I'm not sure I have a complete explanation of why it feels to me this would be confusing to some users, especially beginner users, but here are some thoughts.

I think you could have just as well added a `--into` argument to `jj split` instead of adding `-d`/`-A`/`-B` to `jj squash`. The functionality of the two commands now overlaps significantly, and we don't explain to the users how to reason about this or how to choose.

So, I feel there are two separate UIs here: one with a separate `jj split` and `jj squash` command, and one with a `jj mix` command that does both. One possible end result I imagine is that `jj split` and `jj squash` could become limited versions of `jj mix` commands with different defaults, and we recommend them for beginners.

Another partial answer might be that I'm thinking of `split -r` as the "primary" kind of split, and that meshes particularly badly with the new options you introduced.

> > I wonder if it'd be better to separate the two commands. If we call newsplit the behavior in this PR (we'd clearly need a better name), I think it'd make more sense to have jj newsquah -r a -r b -d c instead of jj newsquah --from a --from b -d c. Also, I think that if a revset has a single root, jj newsquash -r revset should collapse the revset into its root. (I think @martinvonz once suggested it).

> I would definitely go that way for -r too. I've seen it in https://github.com/jj-vcs/jj/issues/5301. But I'm not sure we need a new command for that.

I'm not sure what's best here, but my logic is that I expect `jj split -r revset` to squash the changes into a commit that's not in `revset`, since this is what happens for `jj split -r one_commit`. Whereas my imaginary `jj mix -r revset` could mean `jj mix -r revset -d 'only(roots(revset)-)'` (where `only` is an imaginary command that errors when passed a revset that expands to more than one revision)

**Comment by ilyagr** [comment]
Loosely related: there is a nice `split` interactive GUI concept where you can have some number of commits arranged horizontally with their diffs and you move lines between the diffs left and right (Sapling and VisualJJ have this UI).

I just realized that in such a UI, combining the squash and split commands makes perfect sense; it's just a matter of how many diffs you start with, and how many you end with.

**Comment by glehmann** [comment]
> I think you could have just as well added a `--into` argument to `jj split` instead of adding `-d`/`-A`/`-B` to `jj squash`. The functionality of the two commands now overlaps significantly, and we don't explain to the users how to reason about this or how to choose.
> 

I've read that some would like a `--into` for `jj split`. I'm not sure to see the point, though, as the feature is already covered by `jj squash`, and it would introduce more overlap between the two commands. But I digress.

The point of `-A/-B/-d` for `squash` is to be able to squash *multiple* commits together and land the resulting commit at a specific point in the commit tree.

I agree that the users should be guided to one or another command based on their use case.

The way I see it is that `jj split` takes a single commit as input and produces multiple ones. `jj squash` takes multiple commits as inputs and produces a single output commit, but it obviously doesn't cover well the interactive changes selection of the fileset feature for `jj squash`. This is probably not good enough.

> So, I feel there are two separate UIs here: one with a separate `jj split` and `jj squash` command, and one with a `jj mix` command that does both. One possible end result I imagine is that `jj split` and `jj squash` could become limited versions of `jj mix` commands with different defaults, and we recommend them for beginners.

Do you think (as I do) that `jj split` could be splitting the changes in more than two revisions? I don't think that feature would fit well in `jj mix`.
`jj mix` would then be a kind of super `jj squash`. But why have both?
BTW, if I didn't have any prior experience with git, I would find `jj mix` much clearer than `jj squash`.

> Another partial answer might be that I'm thinking of `split -r` as the "primary" kind of split, and that meshes particularly badly with the new options you introduced.

I think it works well for `jj split`. Do you think they are a problem in `jj split`?

If `jj squash -r` is also the primary kind of squash for you, I can understand that the new `-A/-B/-d` flags don't work for you. But the `--from` was already there, and `-A/-B/-d` work well with it IMO.
In fact `jj squash` already has two behaviors, `--revision` and `--from`. It shows in the clap configuration: `-f/-t/-A/-B/-d` all conflict with `-r`.

> > > I wonder if it'd be better to separate the two commands. If we call newsplit the behavior in this PR (we'd clearly need a better name), I think it'd make more sense to have jj newsquah -r a -r b -d c instead of jj newsquah --from a --from b -d c. Also, I think that if a revset has a single root, jj newsquash -r revset should collapse the revset into its root. (I think @martinvonz once suggested it).
> 
> > I would definitely go that way for -r too. I've seen it in #5301. But I'm not sure we need a new command for that.
> 
> I'm not sure what's best here, but my logic is that I expect `jj split -r revset` to squash the changes into a commit that's not in `revset`, since this is what happens for `jj split -r one_commit`. Whereas my imaginary `jj mix -r revset` could mean `jj mix -r revset -d 'only(roots(revset)-)'` (where `only` is an imaginary command that errors when passed a revset that expands to more than one revision)

I'm not sure, unlike in `jj split`, that the current behavior of `jj squash -r` is actually useful for something else than `@`.
`jj squash`, with no parameter, is definitely useful, but it is equivalent to:
* `jj squash -r @` with the current `jj squash` behavior;
* `jj squash -f @ -t @-`;
* and `jj squash -r @-::@` with the behavior from #5301.

So I'm not sure we need the current behavior of `-r`. Is it worth having it as an equivalent of `-r` in `jj split`? But if you see it as the primary operation, I understand it doesn't work for you.

If we acknowledge that `jj squash` has two modes, as noted earlier (`--revision` and `--from`), and that the current `--revision` is a subset of the new behavior described in #5301, we could move the `--from` behavior in a new `jj mix` command and keep `--revision` in `jj squash`, with the new behavior described in #5301.
`jj squash` would be changed to only implement the new behavior described in #5301, and make `-r` default to `@-::@`. `jj squash`, without argument, would still move the changes from the working copy to the parent, and we would gain an easy way to squash a bunch of commits into the root one.
The current `jj squash --from` would be moved to `jj mix -r`, with the `-A/-B/-d` flags that wouldn't make any sense in `jj squash` in that context.

**Comment by glehmann** [comment]
I'm re-reading my previous message, and `jj mix` feels kind of weird to me: it looks like a squash with another name, and not exactly the same features, but still a squash. And I'm not sure how a new command would make things clearer for a new comer.

If I was to make a `squash` from scratch, I would make:
* `jj squash -r revset` squashes all the input commits in `only(roots(revset))`
* `jj squash` default to `jj squash -r @-::@`, to keep the current default behavior
* `-i` and the fileset to select a part of the changes
* `-A/-B/-d/-t` to choose where to put the resulting changes, if we don't want to have the result in the root of the input commits. `-t` is a bit special, as it also serve as input, but that seems ok to me.

No `--from` would make a bit harder to squash some changes in the current working copy, but `jj squash -r x -t @` is not horrible either.

I stop here. Sorry to add to an already too long message!

**Comment by ilyagr** [comment]
This post is mostly a continuation of my brainstorming, it doesn't really respond to the more concrete parts of your comment. Hope it's helpful for us to get on the same page, but feel free to skip parts of it; it's all conjectural. A part of the reason I wrote it is to record my own brainstorming for my future self.

> The point of -A/-B/-d for squash is to be able to squash multiple commits together and land the resulting commit at a specific point in the commit tree.

I think the case of a `squash -i --from -A/-B/-d` with a single `--from` is also important (especially so for people trying to understand `squash -A/-B/-d`, but also in general).

Split currently covers the "1 commit -> 2 commits" case, while `squash -i -A/-B/-D` now covers the "n commits -> (n+1) commits" case, including n=1. Previously, the two commands were almost completely orthogonal.

I don't currently see `squash` encroaching on `split`'s territory as any worse (or better) than adding `split --into`. Doing one and not the other seems a bit inconsistent, even.

However, I'm not suggesting that we should make both commands do everything, nor am I suggesting that this encroachment is terrible, nor is it urgent to resolve (especially now that the new features are marked as experimental). I'm mostly wondering whether we can reorganize the UI in a way where the commands are more orthogonal and, more importantly, easier to think about. E.g., one command being expressed as a special case of another can be easier to think about than if they merely overlapped.

At the same time, your new features also allow the user to do things they couldn't quite do before. If people find these abilities useful, we'd still want the new UI to continue to be powerful. It's also possible that people find some of the new abilities useful, but not others. Then, we could only keep the useful ones in whatever the new UI will be (and whenever it will be).

So, again, I think it's great to experiment with these features now. The UI you set up is more than good enough for that, and experimenting with it could be crucial to get ideas on what a new UI could be.



> >    Another partial answer might be that I'm thinking of split -r as the "primary" kind of split, and that meshes particularly badly with the new options you introduced.
>
> I think it works well for jj split. Do you think they are a problem in jj split?

That remark of mine wasn't something I thought about very carefully, and certainly wasn't meant to be actionable or to indicate that something is wrong. It was mostly me trying to guess where some of my intuition might be coming from. If it wasn't helpful, feel free to just ignore it.

> Do you think (as I do) that jj split could be splitting the changes in more than two revisions? I don't think that feature would fit well in jj mix.
> jj mix would then be a kind of super jj squash. But why have both?
BTW, if I didn't have any prior experience with git, I would find jj mix much clearer than jj squash.

Again, my original remark was mostly brainstorming. But what you are saying seems pretty close to what I was thinking; I'm also wondering whether `jj mix` could be some sort of a general "n commits" -> "m commits" command.

OK, I do have one thought I believe moderately strongly in: if we have a general `jj mix` command (possibly under another name), it might be difficult to design it to be immediately intuitive to new people. That's why I thought we might still need something like `jj split` to make the common special case of `jj mix` easy to understand. But I only believe moderately strongly that these kinds of concerns should be considered as part of designing the new UI; I'm not sure that the end result would be exactly what I thought of now.

**Comment by ilyagr** [comment]
Unfortunately, I don't think I'm personally ready to sign up to immediately figuring out what the new UI should be, and I don't expect it to happen very quickly. Partially, it's because I want to try out your features first, and I also want to wait and see whether other people will get ideas after playing with them. Partially, I expect to be busy with other things in the near future.
